### ISSUE TITLE
What is the best way to use AllenNLP with scikit-learn/others?

### ISSUE THREAD

**Issue by bryant1410** [allenai/allennlp#2900]
I have a codebase that loads several embeddings types, has several neural baselines and also simple ones, so it's pretty convenient to have many config files and just use AllenNLP. However, I wonder how can I run a scikit-learn's SVM baseline with some transformation of word representations as input, for example, running any baseline seamlessly with AllenNLP.

The classic `Trainer` needs a loss (attached to tensor parameters). The NoOp doesn't enter the `forward` method for the training data. There's also the problem of saving the scikit-learn model. Maybe the solution would be a `SklearnTrainer`?

**Comment by kernelmachine** [comment]
I think the easiest way to use AllenNLP with scikit models is to create an independent script that imports the relevant AllenNLP modules you want for preprocessing/input transformation, and leverages scikit for the actual training and persistence (https://scikit-learn.org/stable/modules/model_persistence.html). 

With the AllenNLP Params object (https://github.com/allenai/allennlp/blob/master/allennlp/common/params.py#L173), you can still use config files to define the parameters of your custom scikit script.

**Comment by joelgrus** [comment]
if I had to do this, I would probably create a (simple) ScikitLearnTrainer like you suggest, I don't know that it would necessarily need to go back into the library, though

**Comment by bryant1410** [comment]
> if I had to do this, I would probably create a (simple) ScikitLearnTrainer like you suggest, I don't know that it would necessarily need to go back into the library, though

Don't you think this is a common issue, in a similar way to `NoOpTrainer`? I feel like AllenNLP goes beyond PyTorch, and as a framework, it's quite useful because it orders things, provides modules and sets up experiments in a convenient way. And running PyTorch models for a particular task along with models from scikit-learn sounds like a common use case to me.

**Comment by joelgrus** [comment]
I don't have a good sense for how common it is, as far as I know no one has asked for this before.

one difference is that there's sort of only one way to write a NoOp trainer, where you could imagine adding infinitely many bells and whistles to a sklearn-trainer, so it could easily get out of hand and become a maintenance nightmare, but I guess it would depend in part on what it looked like. 

I can definitely see the appeal; at the same time we're a small team with no deep sklearn expertise and so it's hard to see how we could officially support "use allennlp to train your sklearn models".

**Comment by bryant1410** [comment]
I understand your concerns and they totally make sense to me. I'll think about this.

Maybe it could be like a generic `SingleTimeTrainer` thing, in which `forward` is called once and no `output['loss']` is expected.

**Comment by DeNeutoy** [comment]
This is out of scope, we don't want to support arbitrary model training.
